jlm/util/GraphWriter.hpp

```diff
@@ -0,0 +1,843 @@
+
+  void
+  OutputDotPortId(std::ostream & out) override;
+
+private:
+  InOutNode & Node_;
+};
+
+class OutputPort : public Port
+{
+  friend InOutNode;
+
+  OutputPort(InOutNode & node);
+
+public:
+  ~OutputPort() override = default;
+
+  const char *
+  GetIdPrefix() const override;
+
+  Node &
+  GetNode() override;
+
+  bool
+  CanBeEdgeHead() const override;
+
+  void
+  SetFillColor(std::string color) override;
+
+  void
+  OutputDotPortId(std::ostream & out) override;
+
+private:
+  InOutNode & Node_;
+};
+
+class InOutNode : public Node
+{
+  friend Graph;
+
+  InOutNode(Graph & graph, size_t inputPorts, size_t outputPorts);
+
+public:
+  ~InOutNode() override = default;
+
+  InputPort &
+  CreateInputPort();
+
+  size_t
+  NumInputPorts();
+
+  InputPort &
+  GetInputPort(size_t index);
+
+  OutputPort &
+  CreateOutputPort();
+
+  size_t
+  NumOutputPorts();
+
+  OutputPort &
+  GetOutputPort(size_t index);
+
+  Graph &
+  CreateSubgraph();
+
+  size_t
+  NumSubgraphs();
+
+  Graph &
+  GetSubgraph(size_t index);
+
+  /**
+   * Regular attributes are placed on the dot node itself.
+   * Use this method to set attributes on the HTML table printed when outputting InOutNodes to dot.
+   * The attributes will affect the outmost table with a visible border.
+   * @param name the name of the attribute
+   * @param value the value the attribute should take
+   * @see SetAttribute for setting attributes on the node itself
+   */
+  void
+  SetHtmlTableAttribute(std::string name, std::string value);
+
+  void
+  SetFillColor(std::string color) override;
+
+  void
+  Finalize() override;
+
+  void
+  OutputASCII(std::ostream & out, size_t indent) override;
+
+  void
+  OutputDot(std::ostream & out, size_t indent) override;
+
+  void
+  OutputSubgraphs(std::ostream & out, GraphOutputFormat format, size_t indent) override;
+
+private:
+  // Attributes that need to be placed on the HTML table in the dot output, and not on the node.
+  std::unordered_map<std::string, std::string> HtmlTableAttributes_;
+
+  std::vector<std::unique_ptr<InputPort>> InputPorts_;
+  std::vector<std::unique_ptr<OutputPort>> OutputPorts_;
+  std::vector<Graph *> SubGraphs_;
+};
+
+/**
+ * Node representing a single sink port
+ */
+class ArgumentNode : public Node
+{
+  friend Graph;
+
+  ArgumentNode(Graph & graph);
+
+public:
+  ~ArgumentNode() override = default;
+
+  const char *
+  GetIdPrefix() const override;
+
+  bool
+  CanBeEdgeHead() const override;
+
+  /**
+   * Indicate that the argument node represents a value coming in from a port in another graph
+   * @param outsideSource the Port in the other graph
+   */
+  void
+  SetOutsideSource(Port & outsideSource);
+
+  void
+  OutputASCII(std::ostream & out, size_t indent) override;
+
+private:
+  // Optional reference to a Port outside of this graph from which this argument came
+  Port * OutsideSource_;
+};
+
+/**
+ * Node representing a single source port
+ */
+class ResultNode : public Node
```

Comment:
Arguments are sink ports and results are source ports. In my head, it is the other way around. Where does the discrepancy come from?